stackslib/src/net/inv/nakamoto.rs

```diff
@@ -188,23 +220,30 @@ impl InvGenerator {
     pub(crate) fn get_processed_tenure(
         &mut self,
         chainstate: &StacksChainState,
-        tip_block_id: &StacksBlockId,
+        tip_block_ch: &ConsensusHash,
+        tip_block_bh: &BlockHeaderHash,
         tenure_id_consensus_hash: &ConsensusHash,
     ) -> Result<Option<InvTenureInfo>, NetError> {
-        if self.processed_tenures.get(tip_block_id).is_none() {
+        let tip_block_id = StacksBlockId::new(tip_block_ch, tip_block_bh);
+        if self.processed_tenures.get(&tip_block_id).is_none() {
             // this tip has no known table.
             // does it have an ancestor with a table? If so, then move its ancestor's table to this
             // tip. Otherwise, make a new table.
-            if let Some(ancestor_tip_id) =
-                self.find_ancestor_processed_tenures(chainstate, tip_block_id)?
+            if let Some((ancestor_tip_id, intermediate_tenures)) =
+                self.find_ancestor_processed_tenures(chainstate, &tip_block_id)?
             {
-                let ancestor_tenures = self
+                let mut ancestor_tenures = self
                     .processed_tenures
                     .remove(&ancestor_tip_id)
                     .unwrap_or_else(|| {
                         panic!("FATAL: did not have ancestor tip reported by search");
                     });
 
+                for ch in intermediate_tenures.into_iter() {
+                    ancestor_tenures.remove(&ch);
+                }
+                ancestor_tenures.remove(tip_block_ch);
```

Comment:
Each entry in `ancestor_tenures` is an `InvTenureInfo`, which contains the `tenure_id_consensus_hash` and `parent_tenure_id_consensus_hash` for the given key (i.e. `ch in intermediate_tenures` and `tip_block_ch`).  Removing them is important, because it prevents caching the _absence_ of a tenure after it has arrived (which is the bug that's being fixed here).

Whenever the function `make_tenure_bitvector` is called on a cold cache, the code will walk the tenure history for the given Stacks tip's reward cycle and generate an `Option<InvTenureInfo>` struct for each sortition.  It will be `None` for a given sortition's consensus hash if either (1) there is no sortition, or (2) there are no Stacks blocks for that tenure.  Importantly, condition (2) can change with the arrival of new Stacks blocks, so we need to evict affected entries in `processed_tenures` when that happens.  We weren't doing this.

This code relies on the fact that a change to the value of a `PeerNetwork`'s `StacksTipInfo` instance, which caches the Stacks chain tip, would imply that new blocks have arrived and been processed since the last call to `make_tenure_bitvector`.  This in turn means that some new blocks have been processed, and some entries in `processed_tenures` need to be evicted since they would now incorrectly report a `None` for a tenure of blocks we processed.

The eviction code here works by looking at all of the tenures in-between the highest cached ancestor of the chain tip and the given chain tip, and if there are any, clearing out their associated `Option<InvTenureInfo>` since the tip advanced past that tenure's boundary.

Reply:
> This code relies on the fact that a change to the value of a PeerNetwork's StacksTipInfo instance, which caches the Stacks chain tip, would imply that new blocks have arrived and been processed since the last call to make_tenure_bitvector. This in turn means that some new blocks have been processed, and some entries in processed_tenures need to be evicted since they would now incorrectly report a None for a tenure of blocks we processed.

Ah, okay. This cache maintenance logic needs to be documented in the code comments. 

> The eviction code here works by looking at all of the tenures in-between the highest cached ancestor of the chain tip and the given chain tip, and if there are any, clearing out their associated Option<InvTenureInfo> since the tip advanced past that tenure's boundary.

I think its also important to mention that it must invalidate _inclusively_: i.e., the highest cached ancestor itself must also be invalidated. I think this patch does that, but it needs to be made clear when you document the cache maintenance.

Reply:
> This cache maintenance logic needs to be documented in the code comments.

Can do :+1: 

> I think its also important to mention that it must invalidate inclusively: i.e., the highest cached ancestor itself must also be invalidated. I think this patch does that, but it needs to be made clear when you document the cache maintenance.

This happens.  The code for `find_ancestor_processed_tenures()` is guaranteed to return the given tip's tenure ID in its vector of tenure IDs, if it returns anything at all, since the tenure ID for the tip is unconditionally loaded and pushed to that vector before the ancestor search.